### PR TITLE
feat: GIDK 751 | Prateek | Test | Advance receipt and receipt adjustment from ledger and voucher module

### PR DESCRIPTION
**Pull request by prateek-giddh**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
https://giddha.atlassian.net/browse/GIDK-751
https://giddha.atlassian.net/browse/GIDK-752

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

### CODE REVIEW

File: apps/web-giddh/src/app/ledger/components/updateLedgerEntryPanel/updateLedgerEntryPanel.component.html [hunk 1]
```diff
@@ -783,19 +775,28 @@ <h3 *ngIf="!openDropDown" (dblclick)="openHeaderDropDown()" class="d-inline-bloc
...
-                                    class="icon-edit11 font-10"></i></a>
+                    <div class="clearfix adjust-receipt font-14" *ngIf="isAdjustedWithAdvanceReceipt && (isAdjustAdvanceReceiptSelected)">
+                        <p class="bold">Adjusted Vouchers
+                            <a href="javascript:void(0)" *ngIf="!selectedAdvanceReceiptAdjustInvoiceEditMode" (click)="openAdjustInvoiceEditMode();$event.stopPropagation();">
+                                <i class="icon-edit11 font-10"></i>
+                                <!-- adjusted static code start  -->
+                                <i class="icon-info"></i>
+                                <div class="adusted-info">
+                                    <p class="mb-1"><strong>Adjustment</strong></p>
+                                    <p class="info-text">The selected entry is a part of Invoice# 43242. All the
```
Comment (prateek-giddh) [inline apps/web-giddh/src/app/ledger/components/updateLedgerEntryPanel/updateLedgerEntryPanel.component.html:786]: Currently API changes are undergoing and soon this value will be dynamic

File: apps/web-giddh/src/app/ledger/components/newLedgerEntryPanel/newLedgerEntryPanel.component.ts [hunk 2]
```diff
@@ -1214,16 +1334,40 @@ export class NewLedgerEntryPanelComponent implements OnInit, OnDestroy, OnChange
...
+                customerUniquename: this.currentTxn.selectedAccount ? this.currentTxn.selectedAccount.uniqueName : '',
+                totalTaxableValue: this.currentTxn.total,
+                subTotal: this.currentTxn.total,
+            },
+            accountDetails: {
+                currencySymbol: this.currentTxn.selectedAccount ? this.currentTxn.selectedAccount.currencySymbol : this.blankLedger.baseCurrencyToDisplay.symbol
+            },
+            activeAccountUniqueName: this.activeAccount.uniqueName
+        };
+        console.log('Reached', this.currentTxn);
```
Comment (rohit-walkover) [inline apps/web-giddh/src/app/ledger/components/newLedgerEntryPanel/newLedgerEntryPanel.component.ts:1361]: Remove console

File: apps/web-giddh/src/app/ledger/components/newLedgerEntryPanel/newLedgerEntryPanel.component.ts [hunk 3]
```diff
@@ -1214,16 +1334,40 @@ export class NewLedgerEntryPanelComponent implements OnInit, OnDestroy, OnChange
...
      *
-     * @param {*} event Template ref to get value
+     * @private
      * @memberof NewLedgerEntryPanelComponent
      */
-    public allowAlphanumericChar(event: any): void {
-        if (event && event.value) {
-            this.blankLedger.passportNumber = this.generalService.allowAlphanumericChar(event.value)
-        }
+    private openAdjustPaymentModal() {
```
Comment (rohit-walkover) [inline apps/web-giddh/src/app/ledger/components/newLedgerEntryPanel/newLedgerEntryPanel.component.ts:1370]: Add return type

File: apps/web-giddh/src/app/proforma-invoice/proforma-invoice.component.ts [hunk 4]
```diff
@@ -429,7 +426,7 @@ export class ProformaInvoiceComponent implements OnInit, OnDestroy, AfterViewIni
     /** Inventory Settings */
     public inventorySettings: any;
     public companyCountryCode: string = '';
-    public advanceReceiptAdjustmentData: AdvanceReceiptAdjustment;
+    public advanceReceiptAdjustmentData: VoucherAdjustments;
```
Comment (rohit-walkover) [inline apps/web-giddh/src/app/proforma-invoice/proforma-invoice.component.ts:429]: JS doc

File: apps/web-giddh/src/app/proforma-invoice/proforma-invoice.component.ts [hunk 5]
```diff
@@ -4917,13 +4925,16 @@ export class ProformaInvoiceComponent implements OnInit, OnDestroy, AfterViewIni
...
-     * @param {{ adjustVoucherData: AdvanceReceiptAdjustment, adjustPaymentData: AdjustAdvancePaymentModal }} advanceReceiptsAdjustEvent event that contains advance receipts adjusted data
+     * @param {{ adjustVoucherData: VoucherAdjustments, adjustPaymentData: AdjustAdvancePaymentModal }} advanceReceiptsAdjustEvent event that contains advance receipts adjusted data
      * @memberof ProformaInvoiceComponent
      */
-    public getAdvanceReceiptAdjustData(advanceReceiptsAdjustEvent: { adjustVoucherData: AdvanceReceiptAdjustment, adjustPaymentData: AdjustAdvancePaymentModal }) {
+    public getAdvanceReceiptAdjustData(advanceReceiptsAdjustEvent: { adjustVoucherData: VoucherAdjustments, adjustPaymentData: AdjustAdvancePaymentModal }) {
 
         this.advanceReceiptAdjustmentData = advanceReceiptsAdjustEvent.adjustVoucherData;
         // this.invFormData.voucherDetails.balanceDue = advanceReceiptsAdjustEvent.adjustPaymentData.balanceDue;
+        this.advanceReceiptAdjustmentData.adjustments.forEach(adjustment => {
```
Comment (rohit-walkover) [inline apps/web-giddh/src/app/proforma-invoice/proforma-invoice.component.ts:4935]: nullity check for  "adjustments" array

File: apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts [hunk 6]
```diff
@@ -62,20 +63,25 @@ export class AdvanceReceiptAdjustmentComponent implements OnInit {
...
-    @Output() public closeModelEvent: EventEmitter<boolean> = new EventEmitter(true);
-    @Output() public submitClicked: EventEmitter<{ adjustVoucherData: AdvanceReceiptAdjustment, adjustPaymentData: AdjustAdvancePaymentModal }> = new EventEmitter();
+    @Input() public advanceReceiptAdjustmentUpdatedData: VoucherAdjustments;
+    /** Stores the type of voucher adjustment */
+    @Input() public adjustedVoucherType: AdjustedVoucherType;
+    /** True if the current module is voucher module required as all voucher adjustments are not supported from API */
+    @Input() public isVoucherModule: boolean;
+
+    @Output() public closeModelEvent: EventEmitter<{ adjustVoucherData: VoucherAdjustments, adjustPaymentData: AdjustAdvancePaymentModal }> = new EventEmitter();
+    @Output() public submitClicked: EventEmitter<{ adjustVoucherData: VoucherAdjustments, adjustPaymentData: AdjustAdvancePaymentModal }> = new EventEmitter();
```
Comment (rohit-walkover) [inline apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts:84]: JS doc

File: apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts [hunk 7]
```diff
@@ -106,9 +115,51 @@ export class AdvanceReceiptAdjustmentComponent implements OnInit {
             this.invoiceFormDetails.voucherDetails.tdsTotal = this.invoiceFormDetails.voucherDetails.tdsTotal || 0;
             this.assignVoucherDetails();
         }
-        this.getAllAdvanceReceipts();
+        console.log('Invoice form details: ', this.invoiceFormDetails);
```
Comment (rohit-walkover) [inline apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts:118]: remove console

File: apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts [hunk 8]
```diff
@@ -582,12 +652,45 @@ export class AdvanceReceiptAdjustmentComponent implements OnInit {
...
+     * get adjusted and latter is received when any voucher
+     * is already adjusted and represents the ADJUSTED amount
+     *
+     * @private
+     * @memberof AdvanceReceiptAdjustmentComponent
+     */
+    private formatAdjustmentData(adjustmentData: Array<Adjustment>): void {
+        if (adjustmentData && adjustmentData.length) {
+            adjustmentData.forEach(adjustment => {
+                if (adjustment.adjustmentAmount && adjustment.adjustmentAmount.amountForAccount) {
```
Comment (rohit-walkover) [inline apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts:688]: Please check this condition if  "adjustment.adjustmentAmount.amountForAccount" value will be zero then we not allowing to update balance due

File: apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts [hunk 9]
```diff
@@ -582,12 +652,45 @@ export class AdvanceReceiptAdjustmentComponent implements OnInit {
...
+                    adjustment.balanceDue = adjustment.adjustmentAmount;
+                } else if (adjustment.balanceDue && adjustment.balanceDue.amountForAccount){
```
Comment (rohit-walkover) [inline apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts:690]: Please verify once if adjustment.balanceDue.amountForAccount is zero

File: apps/web-giddh/src/app/ledger/components/newLedgerEntryPanel/newLedgerEntryPanel.component.ts (hunk 2 above)
Comment (prateek-giddh) [reply apps/web-giddh/src/app/ledger/components/newLedgerEntryPanel/newLedgerEntryPanel.component.ts:1361]: Please ignore console for now as mentioned in PR comment @rohit-walkover

File: apps/web-giddh/src/app/ledger/components/newLedgerEntryPanel/newLedgerEntryPanel.component.ts (hunk 3 above)
Comment (prateek-giddh) [reply apps/web-giddh/src/app/ledger/components/newLedgerEntryPanel/newLedgerEntryPanel.component.ts:1370]: Done @rohit-walkover

File: apps/web-giddh/src/app/proforma-invoice/proforma-invoice.component.ts (hunk 4 above)
Comment (prateek-giddh) [reply apps/web-giddh/src/app/proforma-invoice/proforma-invoice.component.ts:429]: Done @rohit-walkover

File: apps/web-giddh/src/app/proforma-invoice/proforma-invoice.component.ts (hunk 5 above)
Comment (prateek-giddh) [reply apps/web-giddh/src/app/proforma-invoice/proforma-invoice.component.ts:4935]: Done @rohit-walkover

File: apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts [hunk 10]
```diff
@@ -106,9 +115,51 @@ export class AdvanceReceiptAdjustmentComponent implements OnInit {
...
+            }
+            this.salesService.getInvoiceList(requestObject, this.invoiceFormDetails.voucherDetails.voucherDate).subscribe((response) => {
+                console.log('Response received: ', response);
+                if (response && response.body) {
+                    this.allAdvanceReceiptResponse = response.body.results
+                    this.adjustVoucherOptions = [];
+                    if (this.allAdvanceReceiptResponse && this.allAdvanceReceiptResponse.length) {
+                        this.allAdvanceReceiptResponse.forEach(item => {
+                            if (item) {
+                                item.voucherDate = item.voucherDate.replace(/-/g, '/');
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/shared/advance-receipt-adjustment/advance-receipt-adjustment.component.ts:148]: if there are any chances of voucherDate being null, add nulity check.

**Pull request merged**